virtual_rainforest/core/data.py

```diff
@@ -0,0 +1,362 @@
+* coerce the data into a properly structured :class:`~xarray.DataArray` object, and then
+* use the :meth:`~virtual_rainforest.core.data.Data.__setitem__` method to validate and
+  add it to a :class:`~virtual_rainforest.core.data.Data` instance.
+
+The  :meth:`~virtual_rainforest.core.data.Data.load_from_file` method implements this
+general recipe for known file formats, with readers described
+:class:`~virtual_rainforest.core.readers` module. See the details of that module for
+supported formats and for extending the system to additional file formats.
+
+Using a data configuration
+--------------------------
+
+A :class:`~virtual_rainforest.core.data.Data` instance can also be populated using the
+:meth:`~virtual_rainforest.core.data.Data.load_from_config` method. This is expecting to
+take a properly validated configuration dictionary, typically loaded from a TOML file
+during configuration (see :class:`~virtual_rainforest.core.config`). The expected
+structure is as follows:
+
+.. code-block:: toml
+
+    [[core.data.variable]]
+    file="/path/to/file.nc"
+    file_var_name="precip"
+    data_var_name="precipitation"
+    [[core.data.variable]]
+    file="/path/to/file.nc"
+    file_var_name="temperature"
+    [[core.data.variable]]
+    file="/path/to/file.csv"
+    file_var_name="elev"
+    data_var_name="elevation"
+
+Note that the properties for each variable in the configuration file are just the
+arguments for :meth:`~virtual_rainforest.core.data.Data.load_from_file`. Data
+configurations must not contain repeated data variable names.  The `data_var_name` is
+optional and is used to change the variable name used in the file to a different value
+to be used within the simulation.
+"""  # noqa: D205
+
+from pathlib import Path
+from typing import Any, Optional
+
+import numpy as np
+from xarray import DataArray, Dataset
+
+from virtual_rainforest.core.axes import AXIS_VALIDATORS, validate_dataarray
+from virtual_rainforest.core.config import ConfigurationError
+from virtual_rainforest.core.grid import Grid
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+from virtual_rainforest.core.readers import FILE_FORMAT_REGISTRY
+
+
+class Data:
+    """The Virtual Rainforest data object.
+
+    This class holds data for a Virtual Rainforest simulation. It functions like a
+    dictionary but the class extends the dictionary methods to provide common methods
+    for data validation etc and to hold key attributes, such as the underlying spatial
+    grid.
+
+    Args:
+        grid: A Grid instance that loaded datasets with spatial structure must match.
+
+    Attrs:
+        grid: The grid instance
+    """
+
+    def __init__(self, grid: Grid) -> None:
+
+        # Set up the instance properties
+        if not isinstance(grid, Grid):
+            log_and_raise("Data must be initialised with a Grid object", TypeError)
+
+        self.grid = grid
+        self.data = Dataset()
```

Comment:
Should `self.data` be documented as an attribute? I'm guessing underscored methods like `_variable_validation` below shouldn't be?

Reply:
I think the private attributes should have a docstring but they won't by default be included in the built docs.